internal/shared/controllers/pull_secret_controller.go

```diff
@@ -0,0 +1,228 @@
+		Named("service-account-controller").
+		WithEventFilter(newNamespacedPredicate(r.ServiceAccountKey)).
+		Build(r)
+
+	return err
+}
+
+// Filters based on the global SecretKey, or any pull secret from the serviceaccount
+func newSecretPredicate(r *PullSecretReconciler) predicate.Predicate {
+	return predicate.NewPredicateFuncs(func(obj client.Object) bool {
+		nn := types.NamespacedName{Namespace: obj.GetNamespace(), Name: obj.GetName()}
+		if r.SecretKey != nil && nn == *r.SecretKey {
+			return true
+		}
+		for _, ps := range r.ServiceAccountPullSecrets {
+			if nn == ps {
+				return true
+			}
+		}
+		return false
+	})
+}
+
+func newNamespacedPredicate(key types.NamespacedName) predicate.Predicate {
+	return predicate.NewPredicateFuncs(func(obj client.Object) bool {
+		return obj.GetName() == key.Name && obj.GetNamespace() == key.Namespace
+	})
+}
+
+// Golang representation of the docker configuration - either dockerconfigjson or dockercfg formats.
+// This allows us to merge the two formats together, regardless of type, and dump it out as a
+// dockerconfigjson for use my contaners/images
+type dockerConfigJSON struct {
+	Auths dockerCfg `json:"auths"`
+}
+
+type dockerCfg map[string]authEntries
+
+type authEntries struct {
+	Auth  string `json:"auth"`
+	Email string `json:"email,omitempty"`
+}
+
+// writeSecretToFile writes the secret data to the specified file
+func (r *PullSecretReconciler) writeSecretToFile(logger logr.Logger, secrets []*corev1.Secret) error {
+	// image registry secrets are always stored with the key .dockerconfigjson or .dockercfg
+	// ref: https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/#registry-secret-existing-credentials
+	// expected format for auth.json
+	// ref: https://github.com/containers/image/blob/main/docs/containers-auth.json.5.md
+
+	jsonData := dockerConfigJSON{}
+	jsonData.Auths = make(dockerCfg)
+
+	for _, s := range secrets {
+		if secretData, ok := s.Data[".dockerconfigjson"]; ok {
+			// process as dockerconfigjson
+			dcj := &dockerConfigJSON{}
+			if err := json.Unmarshal(secretData, dcj); err != nil {
+				return err
+			}
+			for n, v := range dcj.Auths {
+				jsonData.Auths[n] = v
+			}
+			continue
+		}
+		if secretData, ok := s.Data[".dockercfg"]; ok {
+			// process as dockercfg, despite being a map, this has to be Unmarshal'd as a pointer
+			dc := &dockerCfg{}
+			if err := json.Unmarshal(secretData, dc); err != nil {
+				return err
+			}
+			for n, v := range *dc {
+				jsonData.Auths[n] = v
+			}
+			continue
+		}
+		// Ignore the unknown secret
+		logger.Info("expected secret.Data key not found", "secret", types.NamespacedName{Name: s.Name, Namespace: s.Namespace})
+	}
+
+	data, err := json.Marshal(jsonData)
+	if err != nil {
+		return fmt.Errorf("failed to marshal secret data: %w", err)
+	}
+	err = os.WriteFile(r.AuthFilePath, data, 0600)
+	if err != nil {
+		return fmt.Errorf("failed to write secret data to file: %w", err)
+	}
```

Comment:
This could be racy because writing this file could potentially happen at the same time that our other controllers are trying to read it. Maybe use `github.com/google/renameio/v2` to get atomic writes?

Reply:
I'll look into it.